File: compatibility-tests/sdk-test-java/src/test/java/io/floci/gcp/test/KafkaTest.java
package io.floci.gcp.test;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.*;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.List;
import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;

@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
class KafkaTest {

    private static final String PROJECT = TestFixtures.projectId();
    private static final String LOCATION = "us-central1";
    private static final String CLUSTER_ID = TestFixtures.uniqueName("java-cluster");
    private static final String TOPIC_ID = TestFixtures.uniqueName("java-topic");

    private static final HttpClient http = HttpClient.newHttpClient();
    private static final ObjectMapper json = new ObjectMapper();

    private static String base() {
        return TestFixtures.endpoint() + "/v1/projects/" + PROJECT + "/locations/" + LOCATION;
    }

    @AfterAll
    static void tearDown() throws Exception {
        delete("/v1/projects/" + PROJECT + "/locations/" + LOCATION + "/clusters/" + CLUSTER_ID);
    }

    @Test
    @Order(1)
    void createCluster() throws Exception {
        String body = json.writeValueAsString(Map.of(
                "capacityConfig", Map.of("vcpuCount", 3, "memoryBytes", 3221225472L),
                "gcpConfig", Map.of("accessConfig", Map.of("networkConfigs",
                        List.of(Map.of("subnet", "projects/test/regions/us-central1/subnetworks/default"))))));

        JsonNode resp = post("/v1/projects/" + PROJECT + "/locations/" + LOCATION + "/clusters?clusterId=" + CLUSTER_ID, body);

        assertThat(resp.path("done").asBoolean()).isTrue();
        assertThat(resp.path("response").path("name").asText()).contains(CLUSTER_ID);
        assertThat(resp.path("response").path("state").asText()).isEqualTo("ACTIVE");
    }

    @Test
    @Order(2)
    void getCluster() throws Exception {
        JsonNode resp = get("/v1/projects/" + PROJECT + "/locations/" + LOCATION + "/clusters/" + CLUSTER_ID);

        assertThat(resp.path("name").asText()).contains(CLUSTER_ID);
        assertThat(resp.path("state").asText()).isEqualTo("ACTIVE");
        assertThat(resp.path("bootstrapAddress").asText()).isNotBlank();
    }

    @Test
    @Order(3)
    void listClusters() throws Exception {
        JsonNode resp = get("/v1/projects/" + PROJECT + "/locations/" + LOCATION + "/clusters");

        boolean found = false;
        for (JsonNode c : resp.path("clusters")) {
            if (c.path("name").asText().contains(CLUSTER_ID)) {
                found = true;
                break;
            }
        }
        assertThat(found).isTrue();
    }

    @Test
    @Order(4)
    void updateCluster() throws Exception {
        String body = json.writeValueAsString(Map.of("capacityConfig", Map.of("vcpuCount", 6, "memoryBytes", 6442450944L)));
        String path = "/v1/projects/" + PROJECT + "/locations/" + LOCATION + "/clusters/" + CLUSTER_ID;

        JsonNode resp = patch(path, body);

        assertThat(resp.path("done").asBoolean()).isTrue();
        assertThat(resp.path("response").path("vcpuCount").asLong()).isEqualTo(6L);
    }

    @Test
    @Order(5)
    void createTopic() throws Exception {
        String body = json.writeValueAsString(Map.of("partitionCount", 3, "replicationFactor", 1));
        String path = "/v1/projects/" + PROJECT + "/locations/" + LOCATION
                + "/clusters/" + CLUSTER_ID + "/topics?topicId=" + TOPIC_ID;

        JsonNode resp = post(path, body);

        assertThat(resp.path("name").asText()).contains(TOPIC_ID);
        assertThat(resp.path("partitionCount").asInt()).isEqualTo(3);
    }

    @Test
    @Order(6)
    void getTopic() throws Exception {
        JsonNode resp = get("/v1/projects/" + PROJECT + "/locations/" + LOCATION
                + "/clusters/" + CLUSTER_ID + "/topics/" + TOPIC_ID);

        assertThat(resp.path("name").asText()).contains(TOPIC_ID);
        assertThat(resp.path("partitionCount").asInt()).isEqualTo(3);
    }

    @Test
    @Order(7)
    void listTopics() throws Exception {
        JsonNode resp = get("/v1/projects/" + PROJECT + "/locations/" + LOCATION
                + "/clusters/" + CLUSTER_ID + "/topics");

        boolean found = false;
        for (JsonNode t : resp.path("topics")) {
            if (t.path("name").asText().contains(TOPIC_ID)) {
                found = true;
                break;
            }
        }
        assertThat(found).isTrue();
    }

    @Test
    @Order(8)
    void updateTopic() throws Exception {
        String body = json.writeValueAsString(Map.of("partitionCount", 6));
        String path = "/v1/projects/" + PROJECT + "/locations/" + LOCATION
                + "/clusters/" + CLUSTER_ID + "/topics/" + TOPIC_ID;

        JsonNode resp = patch(path, body);

        assertThat(resp.path("partitionCount").asInt()).isEqualTo(6);
    }

    @Test
    @Order(9)
    void listConsumerGroups() throws Exception {
        JsonNode resp = get("/v1/projects/" + PROJECT + "/locations/" + LOCATION
                + "/clusters/" + CLUSTER_ID + "/consumerGroups");

        assertThat(resp.has("consumerGroups")).isTrue();
    }

    @Test
    @Order(10)
    void deleteTopic() throws Exception {
        delete("/v1/projects/" + PROJECT + "/locations/" + LOCATION
                + "/clusters/" + CLUSTER_ID + "/topics/" + TOPIC_ID);

        JsonNode resp = get("/v1/projects/" + PROJECT + "/locations/" + LOCATION
                + "/clusters/" + CLUSTER_ID + "/topics");
        for (JsonNode t : resp.path("topics")) {
            assertThat(t.path("name").asText()).doesNotContain(TOPIC_ID);
        }
    }

    @Test
    @Order(11)
    void deleteCluster() throws Exception {
        delete("/v1/projects/" + PROJECT + "/locations/" + LOCATION + "/clusters/" + CLUSTER_ID);

        JsonNode resp = get("/v1/projects/" + PROJECT + "/locations/" + LOCATION + "/clusters");
        for (JsonNode c : resp.path("clusters")) {
            assertThat(c.path("name").asText()).doesNotContain(CLUSTER_ID);
        }
    }

    // ── HTTP helpers ──────────────────────────────────────────────────────────

    private static JsonNode patch(String path, String body) throws Exception {
        HttpRequest req = HttpRequest.newBuilder()
                .uri(URI.create(TestFixtures.endpoint() + path))
                .header("Content-Type", "application/json")
                .method("PATCH", HttpRequest.BodyPublishers.ofString(body))
                .build();
        HttpResponse<String> resp = http.send(req, HttpResponse.BodyHandlers.ofString());
        assertThat(resp.statusCode()).isEqualTo(200);
        return json.readTree(resp.body());
    }

    private static JsonNode post(String path, String body) throws Exception {
        HttpRequest req = HttpRequest.newBuilder()
                .uri(URI.create(TestFixtures.endpoint() + path))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(body))
                .build();
        HttpResponse<String> resp = http.send(req, HttpResponse.BodyHandlers.ofString());
        assertThat(resp.statusCode()).isEqualTo(200);
        return json.readTree(resp.body());
    }

    private static JsonNode get(String path) throws Exception {
        HttpRequest req = HttpRequest.newBuilder()
                .uri(URI.create(TestFixtures.endpoint() + path))
                .GET()
                .build();
        HttpResponse<String> resp = http.send(req, HttpResponse.BodyHandlers.ofString());
        assertThat(resp.statusCode()).isEqualTo(200);
        return json.readTree(resp.body());
    }

    private static void delete(String path) throws Exception {
        if (path.contains("/null")) {
            return;
        }
        HttpRequest req = HttpRequest.newBuilder()
                .uri(URI.create(TestFixtures.endpoint() + path))
                .DELETE()
                .build();
        http.send(req, HttpResponse.BodyHandlers.ofString());
    }
}

File: compatibility-tests/sdk-test-node/tests/kafka.test.ts
import { describe, it, expect, beforeAll, afterAll } from 'vitest';
import { ENDPOINT, PROJECT_ID, uniqueName } from './setup';

const base = () =>
  `${ENDPOINT}/v1/projects/${PROJECT_ID}/locations/us-central1`;

async function post(url: string, body: unknown = {}): Promise<unknown> {
  const resp = await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });
  expect(resp.status).toBe(200);
  return resp.json();
}

async function get(url: string): Promise<unknown> {
  const resp = await fetch(url);
  expect(resp.status).toBe(200);
  return resp.json();
}

async function patch(url: string, body: unknown = {}): Promise<unknown> {
  const resp = await fetch(url, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });
  expect(resp.status).toBe(200);
  return resp.json();
}

async function del(url: string): Promise<void> {
  await fetch(url, { method: 'DELETE' });
}

describe('Managed Kafka', () => {
  let clusterId: string;
  let topicId: string;

  beforeAll(() => {
    clusterId = uniqueName('node-cluster');
    topicId = uniqueName('node-topic');
  });

  afterAll(async () => {
    if (clusterId) {
      await del(`${base()}/clusters/${clusterId}`);
    }
  });

  it('should create a cluster', async () => {
    const resp = await post(`${base()}/clusters?clusterId=${clusterId}`, {
      capacityConfig: { vcpuCount: 3, memoryBytes: 3221225472 },
      gcpConfig: { accessConfig: { networkConfigs: [] } },
    }) as { done: boolean; response: Record<string, string> };

    expect(resp.done).toBe(true);
    expect(resp.response.name).toContain(clusterId);
    expect(resp.response.state).toBe('ACTIVE');
  });

  it('should get a cluster', async () => {
    const resp = await get(`${base()}/clusters/${clusterId}`) as Record<string, string>;
    expect(resp.name).toContain(clusterId);
    expect(resp.state).toBe('ACTIVE');
    expect(resp.bootstrapAddress).toBeTruthy();
  });

  it('should list clusters', async () => {
    const resp = await get(`${base()}/clusters`) as { clusters: Array<{ name: string }> };
    expect(resp.clusters.some((c) => c.name.includes(clusterId))).toBe(true);
  });

  it('should update a cluster', async () => {
    const resp = await patch(`${base()}/clusters/${clusterId}`, {
      capacityConfig: { vcpuCount: 6, memoryBytes: 6442450944 },
    }) as { done: boolean; response: Record<string, unknown> };
    expect(resp.done).toBe(true);
    expect(resp.response.vcpuCount).toBe(6);
  });

  it('should create a topic', async () => {
    const resp = await post(
      `${base()}/clusters/${clusterId}/topics?topicId=${topicId}`,
      { partitionCount: 3, replicationFactor: 1 },
    ) as Record<string, unknown>;

    expect((resp.name as string)).toContain(topicId);
    expect(resp.partitionCount).toBe(3);
  });

  it('should get a topic', async () => {
    const resp = await get(
      `${base()}/clusters/${clusterId}/topics/${topicId}`,
    ) as Record<string, unknown>;
    expect((resp.name as string)).toContain(topicId);
    expect(resp.partitionCount).toBe(3);
  });

  it('should list topics', async () => {
    const resp = await get(
      `${base()}/clusters/${clusterId}/topics`,
    ) as { topics: Array<{ name: string }> };
    expect(resp.topics.some((t) => t.name.includes(topicId))).toBe(true);
  });

  it('should update a topic', async () => {
    const resp = await patch(
      `${base()}/clusters/${clusterId}/topics/${topicId}`,
      { partitionCount: 6 },
    ) as Record<string, unknown>;
    expect(resp.partitionCount).toBe(6);
  });

  it('should list consumer groups', async () => {
    const resp = await get(
      `${base()}/clusters/${clusterId}/consumerGroups`,
    ) as Record<string, unknown>;
    expect(resp).toHaveProperty('consumerGroups');
  });

  it('should delete topic', async () => {
    await del(`${base()}/clusters/${clusterId}/topics/${topicId}`);

    const resp = await get(
      `${base()}/clusters/${clusterId}/topics`,
    ) as { topics: Array<{ name: string }> };
    expect(resp.topics.every((t) => !t.name.includes(topicId))).toBe(true);
    topicId = '';
  });

  it('should delete cluster', async () => {
    await del(`${base()}/clusters/${clusterId}`);

    const resp = await get(`${base()}/clusters`) as { clusters: Array<{ name: string }> };
    expect(resp.clusters.every((c) => !c.name.includes(clusterId))).toBe(true);
    clusterId = '';
  });
});
